ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_service.proto

```diff
@@ -128,6 +135,74 @@ message GetFlatTransactionResponse {
   Transaction transaction = 1;
 }
 
+message GetEventsByContractIdRequest {
+  // The contract id being queried.
+  // Required
+  string contract_id = 1;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract id will only be returned if the requesting parties includes
+  // at least one party that is a witness of the action that created the event. For a full description
+  // of witness see https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#projections.
+  // Required
+  repeated string requesting_parties = 2;
+
+}
+message GetEventsByContractIdResponse {
+  // The collection of events, will only contain CreateEvent and consuming ExerciseEvents. If the
+  // requested `contract_id` has been archived or never existed this collection will be empty.
+  repeated TreeEvent events = 1;
+}
+
+message GetEventsByContractKeyRequest {
+
+  // The contract key to search for.
+  // Required
+  Value contract_key = 1;
+
+  // The template id associated with the contract key
+  // Required
+  Identifier template_id = 2;
+
+  // The parties whose events the client expects to see.
+  // The events associated with the contract key will only be returned if the requesting parties includes
+  // at least one party that is a witness of the action that created the event. For a full description
+  // of witness see https://docs.daml.com/concepts/ledger-model/ledger-privacy.html#projections.
+  // Required
+  repeated string requesting_parties = 3;
+
+  // Optional - will default to 1000.
+  uint32 max_events = 4;
+
+  // Search from ledger offset after the provided begin_exclusive
+  // Optional - will default to the searching of ledger from the start
+  LedgerOffset begin_exclusive = 5;
+
+  // If provided only return events with an offset <= `end_inclusive`
+  // Optional
+  LedgerOffset end_inclusive = 6;
+
+}
+
+message GetEventsByContractKeyResponse {
+
+  // The collection of events, will only contain CreateEvents and consuming ExerciseEvents. Note
+  // that the number of events contained should not be used to determine if another call is
+  // required to get all rows, use the `last_offset` below for this determination.
+  repeated TreeEvent events = 1;
+
+  // If populated there are potentially more rows, to avoid gaps make
+  // the next request with `begin_exclusive` set to `last_offset`
+  LedgerOffset last_offset = 2;
```

Comment:
From the implementation it seems this is not populated if the unfiltered list is smaller than the limit. this has an uncomfortable consequence: if someone polls the ledger end, if this is not filled, he or she needs to take it from the result list, since the "previous" last offset wont change.
Anyhow this problem bolis down to what it means we query until the ledger end - this means a moving target.
(this kind of problems raise since we are introducing pagination, I would suggest to simplify the query pattern if we can, and most importantly moving this to admin)

Reply:
> is this the last offset from the events above, or the last offset of the queried range
If populated it is the value that should be fed into the next begin_exclusive value (in practice it is the final offset value where the num of rows being returned is the maximum)